### PR TITLE
WIP | BSP-217: JS Common extract & packages creation

### PR DESCRIPTION
**Pull request by 0x103**
- Package for open-api
- Package for nats-runner
- Package for logger
- Package for event-bus

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    "JWA",
+    "LAB",
+    "js-common",
+    "Nats"
+  ],
+  "workspaces": [
+    "./packages/*"
+  ],
+  "devDependencies": {
+    "@jwalab/minilab": "^0.0.15",
```
Comment (tkjwa) [inline package.json]: why is it needed for dev ?

File: packages/event-bus/.npmignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+*
+!dist/*
+!package.json
+!readme.md
```
Comment (tkjwa) [inline packages/event-bus/.npmignore]: missing new line

File: packages/event-bus/.prettierrc.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (tkjwa) [inline packages/event-bus/.prettierrc.json:1]: what do you think about adding `"printWidth": 120,`?

File: packages/event-bus/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export * from "./EventEmitterEventBus";
+export * from "./EventBus";
```
Comment (tkjwa) [inline packages/event-bus/src/index.ts]: missing new line

File: packages/logger/.npmignore [hunk 5]
```diff
@@ -0,0 +1,4 @@
+*
+!dist/*
+!package.json
+!readme.md
```
Comment (tkjwa) [inline packages/logger/.npmignore]: missing new line

File: packages/logger/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+export * from "./Logger";
+export * from "./winstonLoggerFactory";
```
Comment (tkjwa) [inline packages/logger/src/index.ts]: missing new line

File: packages/logger/package.json [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    "@typescript-eslint/eslint-plugin": "^4.4",
+    "@typescript-eslint/parser": "^4.4",
+    "eslint": "^7.32.0",
+    "eslint-config-prettier": "^8.3.0",
+    "openapi-types": "^9.3.0",
+    "prettier": "^2.3.2",
+    "typescript": "^4.4"
+  },
+  "dependencies": {
+      "node-dependency-injection": "^2.7.1",
```
Comment (tkjwa) [inline packages/logger/package.json]: not used

File: packages/nats-runner/.npmignore [hunk 8]
```diff
@@ -0,0 +1,4 @@
+*
+!dist/*
+!package.json
+!readme.md
```
Comment (tkjwa) [inline packages/nats-runner/.npmignore]: missing new line

File: packages/nats-runner/.prettierrc.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (tkjwa) [inline packages/nats-runner/.prettierrc.json:1]: what do you think about adding `"printWidth": 120,`?

File: tsconfig.build.json [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  "exclude": [
+    "**/*.test.ts",
+    "**/*.stub.ts",
+    "node_modules",
+    "dist"
+  ],
+  "extends": "@tsconfig/node14/tsconfig.json",
+  "compilerOptions": {
+    "moduleResolution": "node",
+    "noImplicitAny": true,
```
Comment (tkjwa) [inline tsconfig.build.json]: not needed, see node14/tsconfig.json & strict=true https://www.typescriptlang.org/tsconfig#noImplicitAny

File: tsconfig.build.json [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    "removeComments": true,
+    "preserveConstEnums": true,
+    "outDir": "dist",
+    "declaration": true,
+    "sourceMap": true,
+    "esModuleInterop": true,
```
Comment (tkjwa) [inline tsconfig.build.json]: not needed, see node14/tsconfig.json

File: tsconfig.build.json [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    "resolveJsonModule": true,
+    "paths": {
+      "@jwalab/*": ["./packages/*/"]
+    }
+  }
+}
```
Comment (tkjwa) [inline tsconfig.build.json]: missing new line

File: tsconfig.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./tsconfig.build.json",
+  "compilerOptions": {
+    "baseUrl": "./"
+  },
+  "include": [
+    "./packages/*/**.ts"
+  ]
+}
```
Comment (tkjwa) [inline tsconfig.json]: missing new line

File: packages/logger/package.json [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+      "winston": "^3.3.3"
+  },
+  "scripts": {
+    "prettier": "npx prettier --write .",
+    "lint": "eslint ./src/ --ext .ts --max-warnings 0",
+    "clean": "rm -rf ./dist/*",
+    "dev": "npx tsc -w",
+    "build": "npx tsc -p tsconfig.pkg.json"
```
Comment (tkjwa) [inline packages/logger/package.json]: I don't know if it is meaningful to do it here or if should only be on "applications/services" but the use of the `--sourcemap` option is very handy ! see also https://www.npmjs.com/package/source-map-support

File: packages/nats-runner/package.json [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    "typescript": "^4.4"
+  },
+  "dependencies": {
+    "jwt-decode": "^3.1.2",
+    "@jwalab/logger": "latest",
+    "@jwalab/open-api": "latest",
+    "node-dependency-injection": "^2.7.1"
+  },
+  "peerDependencies": {
+    "nats": "^2.1.0"
```
Comment (tkjwa) [inline packages/nats-runner/package.json]: I'm not sure about that ... we can talk about it

File: packages/open-api/.npmignore [hunk 16]
```diff
@@ -0,0 +1,4 @@
+*
+!dist/*
+!package.json
+!readme.md
```
Comment (tkjwa) [inline packages/open-api/.npmignore]: new line

File: packages/open-api/src/index.ts [hunk 17]
```diff
@@ -0,0 +1 @@
+export * from "./openApi";
```
Comment (tkjwa) [inline packages/open-api/src/index.ts]: new line

File: packages/open-api/package.json [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    "@typescript-eslint/eslint-plugin": "^4.4",
+    "@typescript-eslint/parser": "^4.4",
+    "eslint": "^7.32.0",
+    "eslint-config-prettier": "^8.3.0",
+    "openapi-types": "^9.3.0",
+    "prettier": "^2.3.2",
+    "typescript": "^4.4"
+  },
+  "dependencies": {
+    "node-dependency-injection": "^2.7.1"
```
Comment (tkjwa) [inline packages/open-api/package.json]: not needed

**Review by tkjwa — COMMENTED** [review]
Good job ! Could you add a `.editorconfig` at the root to avoid "missing new lines" please ?

File: packages/logger/package.json (hunk 14 above)
Comment (podefr) [reply packages/logger/package.json]: yes we should definitely include source maps 👍

File: packages/nats-runner/package.json (hunk 15 above)
Comment (podefr) [reply packages/nats-runner/package.json]: it's a good idea to not include NATS as part of the dependencies but still tell which version of NATS this package will work with. I think peerDependencies is the best way to let npn know what version of NATS it's expecting. Happy to hear your thoughts

Comment (tkjwa) [reply packages/nats-runner/package.json]: I see the benefit of being able to choose which version you want to work with in a specific service. I see the downside that you have to wonder which version to use and remember to add it yourself in your dependencies.
If we go that way we should do it for other dependencies as well

File: packages/event-bus/.prettierrc.json (hunk 3 above)
Comment (0x103) [reply packages/event-bus/.prettierrc.json:1]: Looks good to me, more human-friendly. Let's update the other projects too then.

File: tsconfig.build.json (hunk 10 above)
Comment (0x103) [reply tsconfig.build.json:11]: ```strict``` is now enabled

File: packages/logger/package.json (hunk 14 above)
Comment (0x103) [reply packages/logger/package.json]: Already enabled in ```tsconfig.build.json``` :)

File: packages/open-api/.npmignore (hunk 16 above)
Comment (0x103) [reply packages/open-api/.npmignore]: Fixed

File: packages/open-api/src/index.ts (hunk 17 above)
Comment (0x103) [reply packages/open-api/src/index.ts]: Fixed

File: packages/open-api/package.json (hunk 18 above)
Comment (0x103) [reply packages/open-api/package.json]: Fixed

File: tsconfig.json (hunk 13 above)
Comment (0x103) [reply tsconfig.json]: Fixed

File: tsconfig.build.json (hunk 12 above)
Comment (0x103) [reply tsconfig.build.json]: Fixed

File: packages/nats-runner/.npmignore (hunk 8 above)
Comment (0x103) [reply packages/nats-runner/.npmignore]: Fixed

File: packages/logger/package.json (hunk 7 above)
Comment (0x103) [reply packages/logger/package.json]: Fixed

File: packages/logger/src/index.ts (hunk 6 above)
Comment (0x103) [reply packages/logger/src/index.ts]: Fixed

File: packages/logger/.npmignore (hunk 5 above)
Comment (0x103) [reply packages/logger/.npmignore]: Fixed

File: packages/event-bus/src/index.ts (hunk 4 above)
Comment (0x103) [reply packages/event-bus/src/index.ts]: Fixed

File: packages/event-bus/.npmignore (hunk 2 above)
Comment (0x103) [reply packages/event-bus/.npmignore]: Fixed

File: package.json (hunk 1 above)
Comment (0x103) [reply package.json]: Fixed

**Review by tkjwa — APPROVED** [review]

**Pull request merged**